src/stirling/bpf_tools/bcc_wrapper.h

```diff
@@ -480,6 +455,176 @@ class BCCWrapper {
+  }
+
+  static std::unique_ptr<WrappedBCCArrayTable> Create(bpf_tools::BCCWrapper* bcc,
+                                                      const std::string& name) {
+    return std::unique_ptr<WrappedBCCArrayTable>(new WrappedBCCArrayTable(bcc, name));
+  }
+
+  StatusOr<T> GetValue(const uint32_t idx) {
+    T value;
+    ebpf::StatusTuple s = underlying_->get_value(idx, value);
+    if (!s.ok()) {
+      char const* const msg = "BCC failed to get value for array table: $0, idx: $1.";
+      return error::Internal(absl::Substitute(msg, name_, idx));
+    }
+    return value;
+  }
+
+  Status SetValue(const uint32_t idx, const T& value) {
+    ebpf::StatusTuple s = underlying_->update_value(idx, value);
+    if (!s.ok()) {
+      return error::Internal(
+          absl::Substitute("BCC failed to set value for array table: $0, idx: $1.", name_, idx));
+    }
+    return Status::OK();
+  }
+
+ private:
+  const std::string name_;
+  std::unique_ptr<U> underlying_;
+};
+
+template <typename K, typename V>
+class WrappedBCCMap {
+ public:
+  using U = ebpf::BPFHashTable<K, V>;
+
+  WrappedBCCMap(bpf_tools::BCCWrapper* bcc, const std::string& name) : name_(name) {
+    underlying_ = std::make_unique<U>(bcc->BPF().get_hash_table<K, V>(name_));
+  }
+
+  static std::unique_ptr<WrappedBCCMap> Create(bpf_tools::BCCWrapper* bcc,
+                                               const std::string& name) {
+    return std::unique_ptr<WrappedBCCMap>(new WrappedBCCMap(bcc, name));
+  }
+
+  size_t capacity() const { return underlying_->capacity(); }
+
+  StatusOr<V> GetValue(const K& key) const {
+    V value;
+    ebpf::StatusTuple s = underlying_->get_value(key, value);
+    if (!s.ok()) {
+      return error::Internal(absl::Substitute("BCC failed to get value for map: $0.", name_));
+    }
+    return value;
+  }
+
+  Status SetValue(const K& key, const V& value) {
+    ebpf::StatusTuple s = underlying_->update_value(key, value);
+    if (!s.ok()) {
+      return error::Internal(
+          absl::Substitute("BCC failed to set value for array table: $0, key: $1.", name_, key));
+    }
+    shadow_keys_.insert(key);
+    return Status::OK();
+  }
+
+  Status RemoveValue(const K& key) {
+    if (shadow_keys_.contains(key)) {
+      const auto s = underlying_->remove_value(key);
+      if (!s.ok()) {
+        return error::Internal(absl::Substitute("BPF failed to remove value for key: $0.", key));
+      }
+      shadow_keys_.erase(key);
+    }
+    return Status::OK();
+  }
+
+  absl::flat_hash_map<K, V> GetTableOffline(const bool clear_table = false) {
+    absl::flat_hash_map<K, V> r;
+
+    for (const auto& k : shadow_keys_) {
+      auto s = GetValue(k);
+      const auto v = s.ConsumeValueOrDie();
+      r[k] = v;
+      if (clear_table) {
+        PX_UNUSED(underlying_->remove_value(k));
+      }
+    }
+    if (clear_table) {
+      shadow_keys_.clear();
+    }
+    return r;
+  }
```

Comment:
Is there a reason this needs to take on reimplementing a function available on the `underlying_` instance? I think it would be best to always defer to the underlying BCC function rather than including BCC implementation details in this shim.

Reply:
We updated the wrapped BCC map to use the shadow keys only if template parameter `kUserSpaceManaged` is set to `true`. It should be the case that iterating over the keys in user space is faster than through iterated syscalls.